helix-view/src/gutter.rs

```diff
@@ -55,36 +57,35 @@ pub fn diagnostic<'doc>(
     let error = theme.get("error");
     let info = theme.get("info");
     let hint = theme.get("hint");
-    let diagnostics = doc.diagnostics();
+    let diagnostics = &doc.diagnostics;
 
     Box::new(
         move |line: usize, _selected: bool, first_visual_line: bool, out: &mut String| {
             if !first_visual_line {
                 return None;
             }
             use helix_core::diagnostic::Severity;
-            if let Ok(index) = diagnostics.binary_search_by_key(&line, |d| d.line) {
-                let after = diagnostics[index..].iter().take_while(|d| d.line == line);
-
-                let before = diagnostics[..index]
-                    .iter()
-                    .rev()
-                    .take_while(|d| d.line == line);
-
-                let diagnostics_on_line = after.chain(before);
-
-                // This unwrap is safe because the iterator cannot be empty as it contains at least the item found by the binary search.
-                let diagnostic = diagnostics_on_line.max_by_key(|d| d.severity).unwrap();
-
-                write!(out, "●").unwrap();
-                return Some(match diagnostic.severity {
+            let first_diag_idx_maybe_on_line = diagnostics.partition_point(|d| d.line < line);
+            if first_diag_idx_maybe_on_line == diagnostics.len() {
+                return None;
+            }
```

Comment:
I don't think you need this, diagnostics[diagnostics.len()..] returns an empty slice so `max_by_key` will return `None` below anyway

Reply:
Yeah good point